### PR TITLE
KNOX-1814 - Moving conf/data folder checking to Java layer from bash

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?
`conf` and `data` directories are configurable using GATEWAY_HOME (or newly introduced env variable) in the Java layer, therefore those checks should either be removed from the `bash` layer, so they take those variables into account within the Java code.

## How was this patch tested?

Running JUnit tests:
```
$ mvn -T1C verify -Prelease,package
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 19:13 min (Wall Clock)
[INFO] Finished at: 2019-03-11T11:04:32+01:00
[INFO] Final Memory: 272M/1965M
[INFO] ------------------------------------------------------------------------
```

Additionally, the following manual test steps were executed:

1. Using non-existing `conf` and `data` folders:
```
$ ls -al $GATEWAY_CONF_HOME
ls: cannot access /home/knox/nonExistingFolder: No such file or directory
$ ls -al $GATEWAY_DATA_HOME
ls: cannot access /home/knox/nonExistingFolder: No such file or directory

$ ./knox-1.3.0-SNAPSHOT/bin/gateway.sh start
Starting Gateway succeeded with PID 32349.

// wait 10 seconds

$ ./knox-1.3.0-SNAPSHOT/bin/gateway.sh status
Gateway is not running. No PID file found.

This error found in gateway.log:
2019-03-11 10:37:13,405 FATAL knox.gateway (GatewayServer.java:main(175)) - Failed to start gateway: org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_DATA_HOME is set to a non-existing directory: /home/knox/nonExistingFolder
GATEWAY_CONF_HOME is set to a non-existing directory: /home/knox/nonExistingFolder
org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_DATA_HOME is set to a non-existing directory: /home/knox/nonExistingFolder
GATEWAY_CONF_HOME is set to a non-existing directory: /home/knox/nonExistingFolder
        at org.apache.knox.gateway.GatewayServer.validateConfigurableGatewayDirectories(GatewayServer.java:260)
        at org.apache.knox.gateway.GatewayServer.main(GatewayServer.java:159)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.knox.gateway.launcher.Invoker.invokeMainMethod(Invoker.java:68)
        at org.apache.knox.gateway.launcher.Invoker.invoke(Invoker.java:39)
        at org.apache.knox.gateway.launcher.Command.run(Command.java:99)
        at org.apache.knox.gateway.launcher.Launcher.run(Launcher.java:75)
        at org.apache.knox.gateway.launcher.Launcher.main(Launcher.java:52)        
```

2. Using non-readable `conf` folder:
```
$ export GATEWAY_CONF_HOME=/home/knox/testFolder
$ mkdir /home/knox/testFolder
$ chmod -r /home/knox/testFolder

$ ./knox-1.3.0-SNAPSHOT/bin/gateway.sh start
Starting Gateway succeeded with PID 11511.

// wait 10 seconds

$ ./knox-1.3.0-SNAPSHOT/bin/gateway.sh status
Gateway is not running. No PID file found.

The following error found in gateway.log:
2019-03-11 10:40:19,813 FATAL knox.gateway (GatewayServer.java:main(175)) - Failed to start gateway: org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_CONF_HOME is set to a non-readable directory: /home/knox/testFolder
org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_CONF_HOME is set to a non-readable directory: /home/knox/testFolder
        at org.apache.knox.gateway.GatewayServer.validateConfigurableGatewayDirectories(GatewayServer.java:260)
        at org.apache.knox.gateway.GatewayServer.main(GatewayServer.java:159)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.knox.gateway.launcher.Invoker.invokeMainMethod(Invoker.java:68)
        at org.apache.knox.gateway.launcher.Invoker.invoke(Invoker.java:39)
        at org.apache.knox.gateway.launcher.Command.run(Command.java:99)
        at org.apache.knox.gateway.launcher.Launcher.run(Launcher.java:75)
        at org.apache.knox.gateway.launcher.Launcher.main(Launcher.java:52)
```

3. Using non-writeable `data` folder:
```
$ export GATEWAY_DATA_HOME=/home/knox/testFolder
$ chmod +r /home/knox/testFolder/
$ chmod -w /home/knox/testFolder/
$ ./knox-1.3.0-SNAPSHOT/bin/gateway.sh start
Starting Gateway failed.

This error found in gateway.log:
2019-03-11 10:42:51,596 FATAL knox.gateway (GatewayServer.java:main(175)) - Failed to start gateway: org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_DATA_HOME is set to a non-writeable directory: /home/knox/testFolder
org.apache.knox.gateway.config.GatewayConfigurationException: Found configurations errors:
GATEWAY_DATA_HOME is set to a non-writeable directory: /home/knox/testFolder
        at org.apache.knox.gateway.GatewayServer.validateConfigurableGatewayDirectories(GatewayServer.java:260)
        at org.apache.knox.gateway.GatewayServer.main(GatewayServer.java:159)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.knox.gateway.launcher.Invoker.invokeMainMethod(Invoker.java:68)
        at org.apache.knox.gateway.launcher.Invoker.invoke(Invoker.java:39)
        at org.apache.knox.gateway.launcher.Command.run(Command.java:99)
        at org.apache.knox.gateway.launcher.Launcher.run(Launcher.java:75)
        at org.apache.knox.gateway.launcher.Launcher.main(Launcher.java:52)
```

### CODE REVIEW

**Comment by smolnar82** [comment]
@risdenk @rlevas @moresandeep @pzampino 
Could any of you please review my changes? Thanks!

**Comment by risdenk** [comment]
@smolnar82 what happens previously with a misconfigured directory?

```bash
./knox-1.3.0-SNAPSHOT/bin/gateway.sh start
Starting Gateway succeeded with PID 32349.
```

I would kind of expect that the gateway wouldn't even say start was successful.

File: gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java [hunk 1]
```diff
@@ -269,10 +280,28 @@ private static void validateKerberosConfig(GatewayConfig config) throws GatewayC
   }
 
   private static void checkIfFileExistsAndCanBeRead(File fileToBeChecked, String propertyName, Set<String> errors) {
+    checkIfFileExistsAndCanBeRead(fileToBeChecked, propertyName, errors, false);
+  }
+
+  private static void checkIfDirectoryExistsAndCanBeRead(File fileToBeChecked, String propertyName, Set<String> errors) {
```
Comment (risdenk) [inline gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java]: Since these are new methods can we pass in a `Path` instead of a `File`?

**Comment by moresandeep** [comment]
Thanks for the PR @smolnar82 ! couple of things that I think would be good to have:
1. In this patch we are checking whether the folder is readable and writable, can we make sure that the permissions are not world readable (which is the case currently), this would be really great to have, something like PosixFilePermissions, this would help with KNOX-1432
`final Set<PosixFilePermission> permissions = new HashSet<>();
permissions.add(PosixFilePermission.OWNER_READ);
permissions.add(PosixFilePermission.OWNER_WRITE);
// not sure if we need this
//permissions.add(PosixFilePermission.OWNER_EXECUTE);
Files.setPosixFilePermissions(file.toPath(), permissions);` 

2. Also, would be nice to have just one function check for read and write, perhaps use a boolean flag to check if you strictly want to check for one over the other.

**Comment by risdenk** [comment]
I would be careful about trying to change the permissions. Maybe we should WARN if the permissions don't match? Changing permissions on a filesystem from a start script seems really wrong.

**Comment by smolnar82** [comment]
> Thanks for the PR @smolnar82 ! couple of things that I think would be good to have:
> 
> 1. In this patch we are checking whether the folder is readable and writable, can we make sure that the permissions are not world readable (which is the case currently), this would be really great to have, something like PosixFilePermissions, this would help with KNOX-1432
>    `final Set<PosixFilePermission> permissions = new HashSet<>(); permissions.add(PosixFilePermission.OWNER_READ); permissions.add(PosixFilePermission.OWNER_WRITE); // not sure if we need this //permissions.add(PosixFilePermission.OWNER_EXECUTE); Files.setPosixFilePermissions(file.toPath(), permissions);`

The goal of this PR is pure read/write check; I believe it's not a good idea to set any permissions on these directories from the code itself.

> 2. Also, would be nice to have just one function check for read and write, perhaps use a boolean flag to check if you strictly want to check for one over the other.

Will see what can be merged/overloaded (still applying clean-code rules)

**Comment by smolnar82** [comment]
> @smolnar82 what happens previously with a misconfigured directory?
> 
> ```shell
> ./knox-1.3.0-SNAPSHOT/bin/gateway.sh start
> Starting Gateway succeeded with PID 32349.
> ```
> 
> I would kind of expect that the gateway wouldn't even say start was successful.

I agree, but this is how the bash script checks if the service started (as of now). A new JIRA should be filed for an enhanced status check (e.g. not purely relying on the process being created and a PID is associated). But this is outside of the scope of this JIRA. Any objection?

**Comment by smolnar82** [comment]
@risdenk @moresandeep 
I improved the private methods to use 'Path` and to have only one method which actually executes the check logic (however I kept the _named_ private methods to make the code easier to read and understand)

File: gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java (hunk 1 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java]: Fixed

**Comment by lfrancke** [comment]
Unfortunately I didn't have time to follow through on this but maybe this discussion from last October or so helps a bit: https://lists.apache.org/thread.html/44d3a2dd8a6fb40baa0120acc88c9244404653706974fb422cb7d546@%3Cdev.knox.apache.org%3E as well as this https://github.com/lfrancke/knox-config-docs

It's great to see this happening now, thank you @smolnar82 ! 
All it took was a huge merger :)

**Comment by smolnar82** [comment]
> Unfortunately I didn't have time to follow through on this but maybe this discussion from last October or so helps a bit: https://lists.apache.org/thread.html/44d3a2dd8a6fb40baa0120acc88c9244404653706974fb422cb7d546@%3Cdev.knox.apache.org%3E as well as this https://github.com/lfrancke/knox-config-docs
> 
> It's great to see this happening now, thank you @smolnar82 !
> All it took was a huge merger :)

Hehe :) The merger helped me to be part of this fantastic team :)
You may check out the other PRs too: #65, #67 

Bash script refinement is on the way.

I read that e-mail thread and even used your Github project to integrate Knox 1.2.0 with CM; thanks for that work!

**Comment by risdenk** [comment]
> I agree, but this is how the bash script checks if the service started (as of now).

@smolnar82 are you sure about this? I can double check in a bit, but it seems like if the check is in the bash script, the script would fail BEFORE java is even started. This would mean that there is no success message printed?

**Comment by smolnar82** [comment]
> > I agree, but this is how the bash script checks if the service started (as of now).
> 
> @smolnar82 are you sure about this? I can double check in a bit, but it seems like if the check is in the bash script, the script would fail BEFORE java is even started. This would mean that there is no success message printed?

Yes, I'm.
If the folder check is moved to Java layer then the `appStart` function will execute the Java process and will chek for the PID file being created; once it finds it considers the startup process finished with success. It may happen that the Java process did not reach the validation at this time (so that it has not thrown the corresponding exception yet) -> the PID file won't be removed.
An easy to implement solution could be to give a bit more time - let's say 2 seconds - to the Java process before the first time PID check (by this time the validation would fail and the Java process would return with 1).
Ideally the check should not rely only on the PID file being created; some sort of heartbeat would be helpful or implementing a basic state machine in the Gateway server (STOPPED, INITIALIZING, STARTED, STOPPING)...but this is out of the scope of this task.
Any idea?

**Comment by smolnar82** [comment]
@risdenk 
Additional information: working on the bash script enhancement JIRA now and adding a 1-second wait before the first PID check solves the issue.

**Pull request merged**